src/PointwiseFunctions/Elasticity/ConstitutiveRelations/Tags.hpp

```diff
@@ -3,37 +3,54 @@
 
 #pragma once
 
+#include <cstddef>
+#include <memory>
 #include <string>
 
 #include "DataStructures/DataBox/DataBoxTag.hpp"
-#include "DataStructures/Tensor/TypeAliases.hpp"
 #include "PointwiseFunctions/Elasticity/ConstitutiveRelations/ConstitutiveRelation.hpp"
+#include "Utilities/TMPL.hpp"
 
-/// \cond
 namespace Elasticity {
 namespace Tags {
-template <size_t Dim>
-struct Stress;
-}  // namespace Tags
-}  // namespace Elasticity
-namespace Tags {
-template <size_t Dim, typename Frame>
-struct Coordinates;
-}  // namespace Tags
-/// \endcond
-
-namespace Elasticity {
-namespace Tags {
-
-/// Base tag for the constitutive relation
-struct ConstitutiveRelationBase : db::BaseTag {};
 
-template <typename ConstitutiveRelationType>
-struct ConstitutiveRelation : ConstitutiveRelationBase, db::SimpleTag {
-  static constexpr OptionString help = {
-      "The constitutive relation of the elastic material"};
-  using type = ConstitutiveRelationType;
+/*!
+ * \brief The elastic material's constitutive relation.
+ *
+ * \see `Elasticity::ConstitutiveRelations::ConstitutiveRelation`
+ */
+template <size_t Dim>
+struct ConstitutiveRelation : db::SimpleTag {
   static std::string name() noexcept { return "Material"; }
+  using type =
+      std::unique_ptr<ConstitutiveRelations::ConstitutiveRelation<Dim>>;
+};
+
+/*!
+ * \brief The elastic material's constitutive relation, loaded from options in
+ * the `ProviderOptionTag`.
+ *
+ * Retrieves the constitutive relation from the object constructed from the
+ * `ProviderOptionTag` by calling its `constitutive_relation()` member function.
+ * Also requires the `ProviderOptionTag::type` to provide a `volume_dim` static
+ * variable and a `constitutive_relation_type` type alias.
+ *
+ * The constitutive relation can be retrieved from the DataBox using its base
+ * `Elasticity::Tags::ConstitutiveRelation` tag.
+ */
+template <typename ProviderOptionTag,
+          typename ProviderType = tmpl::type_from<ProviderOptionTag>>
+struct ConstitutiveRelationFrom
+    : ConstitutiveRelation<ProviderType::volume_dim> {
+  using base = ConstitutiveRelation<ProviderType::volume_dim>;
+  using type = tmpl::type_from<base>;
+  using option_tags = tmpl::list<ProviderOptionTag>;
+
+  template <typename Metavariables>
+  static type create_from_options(const ProviderType& provider) {
+    return std::make_unique<typename ProviderType::constitutive_relation_type>(
+        provider.constitutive_relation());
+  }
 };
```

Comment:
What kind of metavariables have you had in mind here? Also, I take it the create_from_options here is the one used below with Parallel::create_from_options<>(), although I don't get how? I have found a Parallelization.md file that I hope will clear things up for me.

Reply:
The `Metavariables` template parameter is part of the compile-time interface that the `Parallel::create_from_options` function expects. Our [dev guide on the parallelization infrastructure](https://spectre-code.org/dev_guide_parallelization_foundations.html) is indeed a good place to start :)

Reply:
since #2026 has been merged, delete the template parameter and add `static constexpr bool pass_metavariables = false;` as we've made it optional to know the metavariables.

Reply:
also why is this a separate tag? why can't you just put `create_from_options` into the previous tag?  if you want them as separate tags, then this one should be the `SimpleTag` and the previous one should be the `BaseTag`

Reply:
- It needs to be a separate tag because it retrieves the constitutive relation from an option-created provider (e.g. the analytic solution) that it is templated on. The `Tags::ConstitutiveRelation` simple tag does not need to know where the constitutive relation it holds comes from; in particular it shouldn't need to be templated on the provider every time it's used to retrieve the constitutive relation from the DataBox
- The `Tags::ConstitutiveRelation` should be a simple tag because it know what type it holds (but is agnostic to where its value came from). It can be used stand-alone to put a constitutive relation in the DataBox if it is not being retrieved from a provide.
This reflects my understanding of DataBox tags, please correct me if I got something wrong

Reply:
did you check that you can retrieve from the DataBox via `ConstitutiveRelation` if it is put into the box with `ConstitutiveRelationFrom`?

Reply:
Yes! I added a test in a fixup commit. However, this is just the standard behavior of our DataBox, so it's probably not necessary to test that for every tag.

Reply:
It seems to me `ConstitutiveRelation` in Elasticity is similar to `EquationOfState` in GRMHD.  You either specify it freely , or you need to grab it from the analytic solution.  If this is the case, please do what is done in the GRMHD executable for an analytic solution

Reply:
Not sure if you still have objections to this @kidder. Here's why I chose the current design:
- We need to retrieve the constitutive relation in the system fluxes, which use an `argument_tags`-style` databox interface. Therefore the argument-tag can't be a type-erased base tag with a strongly typed simple tag in the databox; at least current that's not supported because the flux must build an `apply` function but can't get the type information from the base tag without the full tags list.
- We could template the elasticity system on the constitutive relation as is done in GRMHD. Or we store the constitutive relation in a `std::unique_ptr` to its base class. I chose the latter because we only need the base class interface when using it. Is there an advantage to the former?

Reply:
@kidder I made a mistake earlier. Turns out there is a way to avoid templating the system on the constitutive relation but keep the constitutive relation tag strongly-typed. This design is almost the same as the EOS tags, with the difference that base tag is typed with the base class `ConstitutiveRelation<Dim>`. This allows us to use the base class in function signatures where the concrete subclass is irrelevant, but it's still possible to retrieve the strongly-typed constitutive relation where the full DataBox tags list is available. The changes reverted a lot of the previous commit's diff, so I pushed a rewritten commit instead of a fixup. Please take a look!

Reply:
@nilsleiffischer I think you need to choose whether the DataBox holds a `std::unique_ptr<ConstitutiveRelation<Dim>>` or a concrete type derived from `ConstitutiveRelation<Dim>`
If you choose the pointer, then I would rename `Tags::ConstitutiveRelationBase` to `Tags::ConstitutiveRelation` and give it the `name` function you want.  If you want to create it directly from the option file, then add an option tag and add the appropriate `create_from_options` (see e.g. `Tags::StepController` from `Time/Tags.hpp`), and put it in the `initialization_tags` of the initialization action; if you want to create it from an analytic solution, the initialization action just grabs it from the solution and puts it into the DataBox.     If you want to have a concrete type, then keep the  `ConstitutiveRelationBase` as a `db::BaseTag` and always retrieve by the base tag (either for `get` or `argument_tags`).  Only the initialization action will need to know the concrete type and you should be able to do this the same way as `EquationOfState` is added in `Evolution/Initialization/ConservativeSystem.hpp`

Reply:
@kidder To help this discussion I opened #2100 that is based on this PR. Let's talk on Monday.

Reply:
This is what I do in #2100:
- The `Tags::ConstitutiveRelationFrom` is placed in the ConstGlobalCache, so it's available through the databox and it's created from the analytic solution through `create_from_options`.
- It can be retrieved either through `Tags::ConstitutiveRelation<ConcreteSubclass>` or `Tags::ConstitutiveRelationBase<Dim>`. With the former you get the object of the concrete subclass type, and with the latter you get it of the virtual base class type. So when the subclass type is not available, you still know it's a `ConstitutiveRelation` and can call the base class functions on it.
- I use the base tag in the fluxes computer, which only needs to know the object is a `ConstitutiveRelation` but doesn't care which subclass.